src/dag.rs

```diff
@@ -202,69 +146,101 @@ impl<G: GraphNode> Dag<G> {
+                .rev(),
+        )
     }
 }
 
+// Possible states of a node during DFS.
+#[derive(Clone, PartialEq, Eq)]
+enum NodeState {
+    New,    // Initial state of every node.
+    Opened, // Node visited (pushed to the stack) but not yet closed.
+    Closed, // Node popped from the stack, all descendants visited.
+}
+
+// Struct that iterates over the nodes reachable by a set of given sources
+// in topological order (https://en.wikipedia.org/wiki/Topological_sorting),
+// that is, every node comes _after_ all its children (yes, techinically
+// this is reverse toposort, but here it kinda makes sense to call it that
+// since edges are meant to represent dependencies).
+//
+// If cycles are present, `next()` will find one eventually and from then
+// on will return `None`.
 #[derive(Clone)]
-pub struct BottomUp<'a, G: GraphNode> {
+pub struct TopologicalSort<'a, G: GraphNode> {
     dag: &'a Dag<G>,
-    visit_stack: Vec<usize>,
-    unvisited_roots: Vec<usize>,
+    dfs_stack: Vec<usize>,
+    node_state: Vec<NodeState>,
+    node_in_cycle: Option<usize>, // Populated when a cycle is detected.
 }
 
-impl<'a, G: GraphNode> Iterator for BottomUp<'a, G> {
-    type Item = &'a G;
-
-    fn next(&mut self) -> Option<&'a G> {
-        // I found the basic non-recursive DFS post-order algorithm here:
-        // https://codingots.medium.com/tree-traversal-without-recursion-221cbea6d004
-        // This is a translation of that, where "s1" is temp_stack and s2 is
-        // self.visit_stack. In that version there is only one root node but
-        // here we have several.
-        // First phase is to build up the stack of nodes to visit using
-        // temp_stack as an intermediate.
-        if self.visit_stack.is_empty() {
-            let mut temp_stack = vec![self.unvisited_roots.pop()?];
-            while let Some(cur_idx) = temp_stack.pop() {
-                self.visit_stack.push(cur_idx);
-                for child_idx in &self.dag.edges[cur_idx] {
-                    temp_stack.push(*child_idx);
-                }
-            }
+impl<'a, G: GraphNode> TopologicalSort<'a, G> {
+    fn new(dag: &'a Dag<G>, sources: Vec<usize>) -> Self {
+        TopologicalSort {
+            dag,
+            dfs_stack: sources,
+            node_state: vec![NodeState::New; dag.nodes.len()],
+            node_in_cycle: None,
         }
-        // Now we just cruise down the to_visit stack drinking a large bottle of
-        // cranberry juice listening to Fleetwood Mac.
-        Some(&self.dag.nodes[self.visit_stack.pop().unwrap()])
     }
-}
 
-#[derive(Clone)]
-pub struct TopDown<'a, G: GraphNode> {
-    dag: &'a Dag<G>,
-    visit_stack: Vec<usize>,
-    // Note these "roots" don't have to be roots in the overall DAG, they are
-    // only roots of the top-down traversals we're doing.
-    unvisited_roots: Vec<usize>,
+    fn check_cycles(mut self) -> Option<usize> {
+        while self.next().is_some() {}
+        self.node_in_cycle
```

Comment:
An idea to avoid the awkward communication via a struct field: `TopologicalSort` could have a separate `next` method in its own `impl` block (I guess you'd have to give it a separate name? Not actually sure how that would work) that returns `Result<Option<Self::Item>, DagError<Self::Item>>`. It would be just like the `Iterator::next` function but it returns `Ok(None)` in the terminal case, `Ok(Some(item))` when returning an item, and `DagError::Cycle` in the current case where it sets `node_in_cycle`.

Then the `Iterator::next` would just be `self.get_next().unwrap()`.

What do you think?

Reply:
Yep good idea, I like this better.